Exit the lock before we call into user code and handle losing the race for the RCW table

jkoritzinsky opened the pull request:
Fixes #110823

jkoritzinsky committed:
Exit the lock before we call into user code and handle losing the race for the RCW table.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
+                        if (!s_rcwTable.TryGetValue(retValue, out NativeObjectWrapper? wrapper))
```
Is the lock around `s_rcwTable.TryGetValue` necessary - can we enter the lock later for `_rcwCache.Add` only?

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
+                                    retValue,
+                                    flags);
+
+                                wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
```
Can this use `TryAdd` like before to avoid the delegate allocation?

jkoritzinsky replied:
`TryAdd` doesn't get us the current value and `TryAdd` + `TryGetValue` would be a race condition (if a GC happened in the middle, we could need to do `TryAdd` again).

I'd like to use `GetValue` that takes a creation delegate, but I'm trying to avoid leaking the unused `NativeObjectWrapper` and relying on the finalizer to clean it up.

jkotas replied:
I think you can do:
```
if (s_rcwTable.TryAdd(retValue, newWrapper))
{
   wrapper = newWrapper;
}
else
{
   wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
}
```
It avoids the delegate allocation for the common case.

Sergio0694 replied:
> "It avoids the delegate allocation for the common case."

@jkotas I remember this also popped up before in an unrelated PR, I don't think doing that is enough to ensure we skip the allocation, unfortunately, See [here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAlAVwDsAXASwFsZCBhCWgB2oBsYoBlbgN2pgYAZwDcAWABQ0jAGY8AJhyNpAb2k4tOVlGr8AhpRh58SHLAMATCOQ4BPFbavUatgxwDqMAwGsAKgbAXAA8EMAAVjBglAA0OOFRMQB8OCIA+lBgAO6BwSYAvDjkMNkAFACUklLaOJra8gmR0ZQ4ALJliS0WMJQAah6kMPFdMcWlnlAGrKzcFfVaGjW12qOt2VMz3NUr2tQAZjhlGVm5QVyE/lD2AIJWVmWw/YPD49mT07NQFfPLu0u7FYbT7cHBFErvTZfHaAgC+CxWMA4IhgCNqAMB2mBWygYLSmRyeQuAHFegMOENHmSXvF0mDUhCPjiqmjtPC/isMAB2HDY6EI9mwoA==), Roslyn has the bad habit of eagerly allocating the display class and using that for locals even if you don't actually need to use the delegate closing over it at runtime. To skip the allocation we need to move the `GetValue` path into a separate small stub method just doing that, to ensure the main block cannot be lowered to use that same display class.

That is, something like this:

```csharp
if (s_rcwTable.TryAdd(retValue, newWrapper))
{
    wrapper = newWrapper;
}
else
{
    static object RcwTableGetValue(object retValue, object newWrapper)
        => s_rcwTable.GetValue(retValue, _ => newWrapper);

    wrapper = RcwTableGetValue(retValue, newWrapper);
}
```

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -1061,17 +1081,19 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                 else
                 {
-                    NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
+                    NativeObjectWrapper newWrapper = NativeObjectWrapper.Create(
```
`_lock` is taken at this spot. Is it a problem?

jkoritzinsky replied:
Yes, good catch. I'll make sure to update this to not hold the lock here.

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                         }
+
                         _rcwCache.Add(identity, wrapper._proxyHandle);
                         AddWrapperToReferenceTrackerHandleCache(wrapper);
```
Trying to understand this better, why is this call needed if we lost the race? In that case, someone else managed to create a new wrapper for the target object. Doesn't that mean that that same someone else would've called `AddWrapperToReferenceTrackerHandleCache` with the same exact wrapper already, meaning doing it here would cause the same wrapper to be registered to the cache multiple times? I can see that cache is basically a custom hash set, meaning it should still work fine, but I'm wondering whether we could get better perf by skipping this call if we lost the race? Or does this not matter because we expect losing the race would be very rare?

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
+                            }
                         }
+
                         _rcwCache.Add(identity, wrapper._proxyHandle);
```
Can you explain why we need to enter the lock again before adding the wrapper to the RCW cache? What I'm not following is: if it's important that we do operations on the RCW table and the RCW cache atomically, then why are we doing `s_rcwTable.GetValue` with the new wrapper **before** entering the lock again? Basically, why don't we enter the lock again right after creating the new wrapper? If instead it's not needed to mutate the RCW table and the RCW cache atomically, then what value does taking the lock specifically just for this one call on the RCW cache provide? What problem is it solving? 🤔

(I guess it also wouldn't hurt to add comments that answer these questions, for future reference)

jkoritzinsky replied:
The RCW cache isn't thread-safe (it's a regular `Dictionary`), so we need to synchronize around editing it. I'll add some comments.

jkoritzinsky committed:
Better encapsulate logic for the RCW table and the RCW cache.

jkoritzinsky committed:
Flip the order of operations to match CoreCLR:

1. Register that we've created a managed object as a wrapper for a COM object first (equivalent to ExtObjCxtCache in CoreCLR)
2. Register the NativeObjectWrapper for lifetime management (equivalent to the sync-block interop info in CoreCLR).

This ensures that we have the same behavior as CoreCLR.

jkoritzinsky committed:
Remove outdated comment

jkoritzinsky commented:
Tests for this specific deadlock are still TODO, but I've reorganized the code for the NativeAOT ComWrappers implementation to match the same order of logic as the CoreCLR implementation and encapsulated each step into separate helper functions (which also helped significantly clarify lock usage).

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -497,9 +494,9 @@ internal unsafe class NativeObjectWrapper
             private IntPtr _inner;
             internal ComWrappers _comWrappers;
-            internal readonly GCHandle _proxyHandle;
             internal readonly GCHandle _proxyHandleTrackingResurrection;
```
I do not think it is an improvement to move ProxyHandle into a property in the middle of the implementation. For types like this, it is important to see the whole state in one place.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -550,15 +556,18 @@ public NativeObjectWrapper(IntPtr externalComObject, IntPtr inner, ComWrappers c
+                if (ProxyHandle.IsAllocated)
                 {
-                    _proxyHandle.Free();
+                    ProxyHandle.Free();
```
This will operate on clone of GCHandle (even before this change). The actual field is going to be a dangling pointer after the Free. Is it a problem?

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -40,17 +40,17 @@ public abstract partial class ComWrappers
+        private static readonly ConditionalWeakTable<object, NativeObjectWrapper> s_nativeObjectWrapperTable = new ConditionalWeakTable<object, NativeObjectWrapper>();
         private static readonly GCHandleSet s_referenceTrackerNativeObjectWrapperCache = new GCHandleSet();
 
         private readonly ConditionalWeakTable<object, ManagedObjectWrapperHolder> _ccwTable = new ConditionalWeakTable<object, ManagedObjectWrapperHolder>();
```
Want to rename this to `_managedObjectWrapperTable` for consistency, since you renamed the other one?

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -40,17 +40,17 @@ public abstract partial class ComWrappers
         private static readonly Guid IID_IWeakReferenceSource = new Guid(0x00000038, 0, 0, 0xC0, 0, 0, 0, 0, 0, 0, 0x46);
 
-        private static readonly ConditionalWeakTable<object, NativeObjectWrapper> s_rcwTable = new ConditionalWeakTable<object, NativeObjectWrapper>();
+        private static readonly ConditionalWeakTable<object, NativeObjectWrapper> s_nativeObjectWrapperTable = new ConditionalWeakTable<object, NativeObjectWrapper>();
```
Should we add a comment to this field and the one below to also mention the fact that "Native object wrapper = RCW" and "Managed object wrapper = CCW"? The acronyms are widely used, and I feel like if we just rename the fields and drop all references to them, other readers might get confused 🤔

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -509,25 +506,34 @@ static NativeObjectWrapper()
             }
 
-            public NativeObjectWrapper(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, CreateObjectFlags flags)
+            internal required GCHandle ProxyHandle { get; init; }
```
+1 to Jan's comment. It feels a bit weird to have fields for everything, and then just this one standalone property somewhere else (on top of the fact this introduces a copy, so you don't mutate the field anymore). If feels just leaving it as a readonly field and having the constructor parameter would be more consistent with the other parameters?

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -1026,60 +1027,88 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
+                    newWrapper.Release();
+                    throw new NotSupportedException();
                 }
                 else
```
Nit: unnecessary `else`, the branch above always throws.

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -1089,14 +1118,54 @@ private static void AddWrapperToReferenceTrackerHandleCache(NativeObjectWrapper
+
+                // Either we added our entry to the cache or
+                // someone else beat us to adding our entry to the cache.
+                // In either case, we want to return the wrapper that one the race.
```
Nit: typo
```suggestion
                // In either case, we want to return the wrapper that won the race.
```

Sergio0694 reviewed (commented):
Thank you for adding all the comments!! 😄
Left a few more notes.

jkoritzinsky committed:
Like CoreCLR, record the NativeObjectWrapper (CoreCLR's ExtObjCxt) in the "RCW address -> wrapper" cache. Use a WeakReference as we don't want to keep the wrapper alive any longer with it being in the cache.

jkoritzinsky committed:
Extract the RCW cache to its own type and do some more cleanup.

jkoritzinsky committed:
Pass the NativeObjectWrapper and comProxy as a pair and ensure that there's only ever one NativeObjectWrapper we try to register, the one that's in the RCW cache

jkoritzinsky committed:
Reorganize code to reduce indentation

jkoritzinsky committed:
A little more refactoring

jkoritzinsky committed:
Rename based on our naming conventions.

jkoritzinsky committed:
Fix UniqueInstance flag

jkoritzinsky committed:
Add a test for the deadlock

jkoritzinsky commented:
Validated that the ComWrappers tests pass locally and added a test for the deadlock. This is ready for review!

jkoritzinsky committed:
Use manually-managed weak GCHandles to reduce allocations.

Co-authored-by: Sergio Pedri <sergio0694@live.com>

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -1089,16 +1103,94 @@ private static void AddWrapperToReferenceTrackerHandleCache(NativeObjectWrapper
         }
 
-        private void RemoveRCWFromCache(IntPtr comPointer, GCHandle expectedValue)
+        private sealed class RcwCache
```
I agree that moving the RCW cache logic into this separate type helps with maintainability, but I'm wondering if we could make this a `struct` (and make the `_rcwCache` field not readonly) to avoid an extra indirection every time we need to use it, which is pretty often? We don't really need this to be a class anyway, it's only ever used internally in this file, and it doesn't have a finalizer or anything. Thoughts? 🙂

AaronRobinsonMSFT replied:
It makes little sense to turn this into a `struct`. Every method here takes a lock. The indirection is noise in most cases and the lock means it won't even be measurable in practice.

AaronRobinsonMSFT commented on src/tests/Interop/COM/ComWrappers/API/Program.cs:
```diff
@@ -962,6 +963,74 @@ public void ValidateAggregationWithReferenceTrackerObject()
+                ppv = IntPtr.Zero;
+                if (iid == ComWrappersHelper.IID_IReferenceTracker && wrappers is not null)
+                {
+                    Console.WriteLine("Attempting to create a new COM object on a different thread.");
```
Is the issue here thread affinity or COM apartment affinity?

jkoritzinsky replied:
I believe the original deadlock was due to COM apartment affinity with the STA vs a Free-threaded apartment thread.

In any case, this test ensures that we're able to create an RCW while we're in the middle of creating an RCW on another thread, which was what was occurring in the original issue.

AaronRobinsonMSFT replied:
My concern here is we've no idea the apartment state in this case. If the apartment state is relevant, then we could be accidentially testing the wrong thing. If this is indeed an STA related issue, then we should create an STA thread and test that as such.

This isn't wrong, just potentially breeding confidence in having more testing coverage than we do.

jkoritzinsky replied:
I've added a test that explicitly marshals a COM object across apartments and hits the right code-paths.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -949,137 +951,149 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
-                    ComInterfaceDispatch* comInterfaceDispatch = TryGetComInterfaceDispatch(identity);
-                    if (comInterfaceDispatch != null)
+                    // If we found a managed object wrapper in this ComWrappers instance
+                    // and it's has the same identity pointer as the one we're creating a NativeObjectWrapper for,
```
```suggestion
                    // and it has the same identity pointer as the one we're creating a NativeObjectWrapper for,
```

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -484,22 +483,20 @@ public ManagedObjectWrapperReleaser(ManagedObjectWrapper* wrapper)
+            private ComWrappers _comWrappers;
+            private GCHandle _proxyHandle;
+            private GCHandle _proxyHandleTrackingResurrection;
             internal readonly bool _aggregatedManagedObjectWrapper;
```
Let's make this a propery instead.

jkoritzinsky committed:
PR feedback

jkoritzinsky committed:
Add a cross-apartment test (Windows-only obviously)

Sergio0694 commented on src/tests/Interop/COM/ComWrappers/API/Program.cs:
```diff
@@ -962,6 +963,172 @@ public void ValidateAggregationWithReferenceTrackerObject()
+                    }
+                    testCompleted.Set();
+                });
+                mtaThread.SetApartmentState(ApartmentState.STA);
```
Did you mean `MTA` here? 😅
```suggestion
                mtaThread.SetApartmentState(ApartmentState.MTA);
```

jkoritzinsky replied:
Yep... Did a few more fixes on the new test as well.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -949,137 +952,149 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
-                if (flags.HasFlag(CreateObjectFlags.Unwrap))
+            // If we have a live cached wrapper currently,
+            // return that.
+            if (_rcwCache.FindProxyForComInstance(identity) is object liveCachedWrapper)
```
Is this also a `null` check?

jkoritzinsky replied:
Yes, this also acts as a null check.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -949,137 +952,149 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
+            if (wrapperMaybe is not null)
+            {
+                retValue = RegisterObjectForComInstance(identity, inner, wrapperMaybe, flags);
+                return true;
```
`return retValue is not null;` was used above, but an explicit `true` is returned here. Is there a difference in these paths?

jkoritzinsky replied:
No there shouldn't be. Good catch. I'll change this one.

jkoritzinsky committed:
PR feedback

AaronRobinsonMSFT commented:
@jkoritzinsky Can you please run GCStress locally? Ideally we'd also include some amount of JITStress too. I'm also fine kicking off the legs that run too.

jkoritzinsky commented:
Do we have GCStress support for NativeAOT?

AaronRobinsonMSFT commented:
@jkoritzinsky Oh, right. I think we do, but we don't have a CI for it. @MichalStrehovsky Do we have a way to run GCStress on local native AOT?

jkotas commented:
There is no GC stress for native AOT.

jkoritzinsky commented:
/azp run runtime-nativeaot-outerloop

Sergio0694 commented:
Not exactly the same (although @EgorBo said it's kinda close enough? 😅), I've added an option in the Microsoft Store to simulate GC stress (I basically just spawn a thread that just spams `GC.Collect(); Thread.Wait(50);` in a loop). Perhaps it'd help if I also tested with that mode enabled for a bit, just to double check things don't blow up when that's running?

jkoritzinsky commented:
NativeAOT outerloop failures are unrelated. Can I get one more review pass?

Sergio0694 commented:
I tested with the Microsoft Store and everything seems to work fine:

- Checked out `release/9.0-staging` (51fd1e2a44e5167170b4aaba823474c2d2e33743)
- Cherry-pick https://github.com/dotnet/runtime/pull/110551
- Cherry-pick https://github.com/dotnet/runtime/pull/110685
- Cherry-pick https://github.com/dotnet/runtime/pull/110828
- Cherry-pick https://github.com/dotnet/runtime/pull/110762
  - Branch with all changes: https://github.com/Sergio0694/runtime/commits/user/sergiopedri/store-testing/
- Build with `.\build.cmd clr.aot -c Release`
- Set `IlcSdkPath` in the Store .csproj to `runtime\artifacts\bin\coreclr\windows.x64.Release\aotsdk\`
- Set `GenerateAppxPackageOnBuild`
- Switch to Release
- Deploy

I used the Store for a while, including closing and launching it a few times, with and without our "GC stress" mode enabled as well (which creates a thread that just spams `GC.Collect()` every 50 milliseconds in a loop). Couldn't find any issues 🎉

Could we backport this one too? I _think_ it should be our last blocker from the NAOT side for shipping 😄
Thank you!
